widgets/heatmap/heatmap.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package heatmap contains a widget that displays heat maps.
+package heatmap
+
+import (
+	"errors"
+	"image"
+	"sync"
+
+	"github.com/mum4k/termdash/cell"
+	"github.com/mum4k/termdash/private/canvas"
+	"github.com/mum4k/termdash/terminal/terminalapi"
+	"github.com/mum4k/termdash/widgetapi"
+	"github.com/mum4k/termdash/widgets/heatmap/internal/axes"
+)
+
+// HeatMap draws heat map charts.
+//
+// Heatmap consists of several cells. Each cell represents a value.
+// The larger the value, the darker the color of the cell (from white to black).
+//
+// The two dimensions of the values (cells) array are determined by the length of
+// the XLabels and YLabels arrays respectively.
+//
+// HeatMap does not support mouse based zoom.
+//
+// Implements widgetapi.Widget. This object is thread-safe.
+type HeatMap struct {
+	// values are the values in the heat map.
+	values [][]float64
+
+	// XLabels are the labels on the X axis in an increasing order.
+	XLabels []string
+	// YLabels are the labels on the Y axis in an increasing order.
+	YLabels []string
+
+	// MinValue and MaxValue are the Min and Max values in the values,
+	// which will be used to calculate the color of each cell.
+	MinValue, MaxValue float64
+
+	// opts are the provided options.
+	opts *options
+
+	// mu protects the HeatMap widget.
+	mu sync.RWMutex
+}
+
+// New returns a new HeatMap widget.
+func New(opts ...Option) (*HeatMap, error) {
+	return nil, errors.New("not implemented")
+}
+
+// Values sets the values to be displayed by the HeatMap.
+// Each value in values has a xLabel and a yLabel, which means
+// len(xLabels) == len(values) and len(yLabels) == len(values[i]).
+// Provided options override values set when New() was called.
+func (hp *HeatMap) Values(xLabels []string, yLabels []string, values [][]float64, opts ...Option) error {
```

Comment:
What you said makes sense, though my design is similar to matlab and python's [1,2].

How about this?  Split the API into three:
``` go
func (hp *HeatMap) Values(values [][]float64, opts ...Option) error { ... }
func (hp *HeatMap) XLabels(xLabels []string, opts ...Option) error { ... }
func (hp *HeatMap) YLabels(yLabels []string, opts ...Option) error { ... }
```
Users can only set data, they don't have to specify labels, then labels will be generated automatically, like `{"0","1","2",...}`.

[1] https://www.mathworks.com/help/matlab/ref/heatmap.html#description
[2] https://seaborn.pydata.org/generated/seaborn.heatmap.html

Reply:
Thanks for providing the API examples. I think your latest proposal is a good one and aligns with the two examples. I think we should also give users a method (or methods) to clear the labels. So they can both add them and remove them.

Are we still going to require that the number of labels matches the number of values on each axis?

One remaining question we should still answer is how should this behave in regards to the limited space available on screen / in the container. Let's say that the user provides values that have 100 data points on the X axis, but the container only provides the HeatMap widget with 50 cells on that axis. What should the widget do?

Reply:
>  I think we should also give users a method (or methods) to clear the labels. So they can both add them and remove them.

It makes sense.

> Are we still going to require that the number of labels matches the number of values on each axis?

I think so. For example, if there are six columns of data, uses have to provide six X labels, but some of them can be empty string.

> One remaining question we should still answer is how should this behave in regards to the limited space available on screen / in the container. Let's say that the user provides values that have 100 data points on the X axis, but the container only provides the HeatMap widget with 50 cells on that axis. What should the widget do?

Sorry, I'm not sure what you mean. Do you mean what should the widget do when the canvas is not big enough to draw all cells of the heat map ? If so, the heat map will not be drawn.

Reply:
Thank you for explaining the behavior when the canvas is not big enough. The proposal sounds good, but we should consider giving the user a hint of how many values they can display. The users of the widget will likely prefer to aggregate the data or push less values, rather than pushing values that will result in nothing displayed. Can we include a function that will hint the user about the available space? There are a couple of examples in the other widgets, e.g. the BarChart's `ValueCapacity()` method:

https://github.com/mum4k/termdash/blob/fbd21e790423e0d479d440e6d1f77e1ca7f01382/widgets/barchart/barchart.go#L241-L256

Lastly after thinking about this widget more - it is also important to give the user a way of atomically updating both the values and the labels. This is because in the default execution mode, `termdash` controls when the screen is redrawn, the user doesn't control it. Now imagine the user wants to display a new set of values with a new set of labels. If we go with separate methods for updating labels and values, the screen could redraw in between causing the terminal to intermittently display inconsistent data.

I think your original proposal of having just one method that accepts the labels and the values was the best one. Sorry that I didn't think about this before. What do you think, should we go back to that proposal?

Reply:
> Can we include a function that will hint the user about the available space?

No problem.

>  What do you think, should we go back to that proposal?

You are right, I didn't think about updating the data either. Let's go back. 😂
Nevertheless, our discussion is still very meaningful and enjoyable, I have benefited a lot. Thank you.